qiskit_experiments/library/driven_freq_tuning/coefficient_utils.py

```diff
@@ -0,0 +1,234 @@
+
+    neg_coef_o3: float
+    """The third order shift coefficient on negative amplitude."""
+
+    offset: float
+    """Offset frequency."""
+
+    def positive_coeffs(self) -> list[float]:
+        """Positive coefficients."""
+        return [self.pos_coef_o3, self.pos_coef_o2, self.pos_coef_o1]
+
+    def negative_coeffs(self) -> list[float]:
+        """Negative coefficients."""
+        return [self.neg_coef_o3, self.neg_coef_o2, self.neg_coef_o1]
+
+    def __str__(self):
+        # Short representation for dataframe
+        return "StarkCoefficients(...)"
+
+
+def convert_freq_to_amp(
+    freqs: np.ndarray,
+    coeffs: StarkCoefficients,
+) -> np.ndarray:
+    """A helper function to convert Stark frequency to amplitude.
+
+    Args:
+        freqs: Target frequency shifts to compute required Stark amplitude.
+        coeffs: Calibrated Stark coefficients.
+
+    Returns:
+        Estimated Stark amplitudes to induce input frequency shifts.
+
+    Raises:
+        ValueError: When amplitude value cannot be solved.
+    """
+    amplitudes = np.zeros_like(freqs)
+    for idx, freq in enumerate(freqs):
+        shift = freq - coeffs.offset
+        if np.isclose(shift, 0.0):
+            amplitudes[idx] = 0.0
+            continue
+        if shift > 0:
+            fit = [*coeffs.positive_coeffs(), -shift]
+        else:
+            fit = [*coeffs.negative_coeffs(), -shift]
+        amp_candidates = np.roots(fit)
+        # Because the fit function is third order, we get three solutions here.
+        # Only one valid solution must exist because we assume
+        # a monotonic trend for Stark shift against tone amplitude in domain of definition.
+        criteria = np.all(
+            [
+                # Frequency shift and tone have the same sign by definition
+                np.sign(amp_candidates.real) == np.sign(shift),
+                # Tone amplitude is a real value
+                np.isclose(amp_candidates.imag, 0.0),
+                # The absolute value of tone amplitude must be less than 1.0 + 10 mp
+                np.abs(amp_candidates.real) < 1.0 + 10 * np.finfo(float).eps,
+            ],
+            axis=0,
+        )
+        valid_amps = amp_candidates[criteria]
+        if len(valid_amps) == 0:
+            raise ValueError(f"Stark shift at frequency value of {freq} Hz is not available.")
+        if len(valid_amps) > 1:
+            # We assume a monotonic trend but sometimes a large third-order term causes
+            # inflection point and inverts the trend in larger amplitudes.
+            # In this case we would have more than one solutions, but we can
+            # take the smallerst amplitude before reaching to the inflection point.
+            before_inflection = np.argmin(np.abs(valid_amps.real))
+            valid_amp = float(valid_amps[before_inflection].real)
+        else:
+            valid_amp = float(valid_amps[0].real)
+        amplitudes[idx] = min(valid_amp, 1.0)
```

Comment:
I coded like you mentioned in the beginning, but found that some test case (round trip case for amplitude input = 1.0) fails due to numerical precision problem. This is the trick to allow full dynamic range more safely.

Reply:
Okay, I see. So you do a sweep up to amp=1.0 and get a set of coefficients. Then you try to sweep up to the max frequency and get an error because rounding causes the max frequency to correspond to an amp slightly above 1 and be filtered out above.

Maybe a comment would be good: "This clipping is okay because values well above 1.0 were filtered out above". I would consider move this operation just below the filtering so it stays close.

By the way, should it be: `np.sign(valid_amp) * min(abs(valid_amp), 1.0)`?